js/core/pubsubhub.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * Module core/pubsubhub
+ *
+ * Returns a singleton that can be used for message broadcasting
+ * and message receiving. Replaces legacy "msg" code in ReSpec.
+ */
+
+"use strict";
+define([], function() {
+  var subscriptions = new Map();
+  var PubSubHub = {
+    pub: function(topic) {
+      // Nothing to do...
+      if (!subscriptions.has(topic)) {
+        return;
+      }
+      var restParams = Array
+        .from(arguments)
+        .slice(1);
+      // we take a copy, because subscribers can unsubscribe themselves
+      // leading to race conditions
+      for (var cb of Array.from(subscriptions.get(topic))) {
+        cb.apply(null, restParams);
+      }
+      // If this is an iframe, postMessage parent (used in testing).
+      if (window.parent === window.self) {
+        return;
+      }
+      var args = restParams.map(function toStructuredClonable(arg) {
+        var str = String(JSON.stringify(arg.stack || arg));
+        return str;
+      });
+      var msg = {
+        topic: topic,
+        args: args,
+      };
+      window.parent.postMessage(msg, "*");
+    },
+    sub: function(topic, cb) {
+      if (!subscriptions.has(topic)) {
+        subscriptions.set(topic, [cb]);
+      } else {
+        subscriptions.get(topic).push(cb);
+      }
+      return {
+        topic: topic,
+        cb: cb,
+      };
+    },
+    unsub: function(opaque) { // opaque is whatever is returned by sub()
+      var callbacks = subscriptions.get(opaque.topic);
+      if (!callbacks || callbacks.indexOf(opaque.cb) === -1) {
+        console.warn("Already unsubscribed:", opaque.topic, opaque.cb);
+        return;
+      }
+      callbacks.splice(callbacks.indexOf(opaque.cb), 1);
+    }
+  };
```

Comment:
I'd be good to have a "seen()" method, which returns a promise that resolves once a message is seen (e.g., "end-all"). That avoids race conditions for one-off events.